File: src-tauri/src/commands/lists.rs
use tauri::State;
use serde_json::{Map, Value};
use crate::models::list::{List, ListField, ListItem, ListSummary};
use crate::services::routing::{agent_rpc, require_agent};
use crate::state::AppState;

#[tauri::command]
pub async fn list_lists(state: State<'_, AppState>) -> Result<Vec<ListSummary>, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/list_lists", serde_json::json!({})).await
}

#[tauri::command]
pub async fn load_list(list_id: String, state: State<'_, AppState>) -> Result<List, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/load_list", serde_json::json!({ "list_id": list_id })).await
}

#[tauri::command]
pub async fn create_list(name: String, fields: Vec<ListField>, state: State<'_, AppState>) -> Result<List, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/create_list", serde_json::json!({ "name": name, "fields": fields })).await
}

#[tauri::command]
pub async fn update_list_meta(list_id: String, name: String, fields: Vec<ListField>, state: State<'_, AppState>) -> Result<List, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/update_list_meta", serde_json::json!({ "list_id": list_id, "name": name, "fields": fields })).await
}

#[tauri::command]
pub async fn delete_list(list_id: String, state: State<'_, AppState>) -> Result<(), String> {
    let url = require_agent(&state).await?;
    let _: serde_json::Value = agent_rpc(&url, "lists/delete_list", serde_json::json!({ "list_id": list_id })).await?;
    Ok(())
}

#[tauri::command]
pub async fn add_list_item(list_id: String, values: Map<String, Value>, state: State<'_, AppState>) -> Result<List, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/add_list_item", serde_json::json!({ "list_id": list_id, "values": values })).await
}

#[tauri::command]
pub async fn update_list_item(list_id: String, item: ListItem, state: State<'_, AppState>) -> Result<List, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/update_list_item", serde_json::json!({ "list_id": list_id, "item": item })).await
}

#[tauri::command]
pub async fn delete_list_item(list_id: String, item_id: String, state: State<'_, AppState>) -> Result<List, String> {
    let url = require_agent(&state).await?;
    agent_rpc(&url, "lists/delete_list_item", serde_json::json!({ "list_id": list_id, "item_id": item_id })).await
}

File: agent/src/tools/lists.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { randomUUID } from "node:crypto";
import { STORAGE_DIR } from "./paths.js";

const LISTS_ROOT = path.join(STORAGE_DIR, "lists");
const LISTS_INDEX = path.join(LISTS_ROOT, "index.json");

export type FieldKind = "text" | "number" | "date" | "bool" | "select";

export interface ListField {
  key: string;
  label: string;
  kind: FieldKind;
  options?: string[] | null;
}

export interface TodoRef {
  date: string;
  todo_id: string;
  label?: string;
}

export interface ListItem {
  id: string;
  values: Record<string, unknown>;
  linked_todos: TodoRef[];
  /** Soft-hide flag. Archived items are kept but hidden from the default view. */
  archived?: boolean;
  created_at: string;
  updated_at: string;
}

export interface List {
  id: string;
  name: string;
  fields: ListField[];
  items: ListItem[];
  created_at: string;
  updated_at: string;
}

export interface ListSummary {
  id: string;
  name: string;
  updated_at: string;
}

function listPath(listId: string): string {
  return path.join(LISTS_ROOT, `${listId}.json`);
}

export function ensureListsDir(): void {
  fs.mkdirSync(LISTS_ROOT, { recursive: true });
}

export function loadListsIndex(): ListSummary[] {
  if (fs.existsSync(LISTS_INDEX)) {
    try {
      return JSON.parse(fs.readFileSync(LISTS_INDEX, "utf-8"));
    } catch {}
  }
  return [];
}

export function saveListsIndex(index: ListSummary[]): void {
  ensureListsDir();
  fs.writeFileSync(LISTS_INDEX, JSON.stringify(index, null, 2), "utf-8");
}

export function loadList(listId: string): List | null {
  const file = listPath(listId);
  if (fs.existsSync(file)) {
    try {
      const raw = JSON.parse(fs.readFileSync(file, "utf-8"));
      raw.fields = raw.fields ?? [];
      raw.items = (raw.items ?? []).map((it: ListItem) => ({
        ...it,
        values: it.values ?? {},
        linked_todos: it.linked_todos ?? [],
        archived: it.archived ?? false,
        created_at: it.created_at || "",
        updated_at: it.updated_at || "",
      }));
      return raw;
    } catch {}
  }
  return null;
}

export function saveList(list: List): void {
  ensureListsDir();
  fs.writeFileSync(listPath(list.id), JSON.stringify(list, null, 2), "utf-8");
}

export function deleteListFile(listId: string): void {
  try {
    fs.unlinkSync(listPath(listId));
  } catch {}
}

function toSummary(list: List): ListSummary {
  return { id: list.id, name: list.name, updated_at: list.updated_at };
}

export function upsertListIndex(list: List): void {
  const index = loadListsIndex();
  const i = index.findIndex((s) => s.id === list.id);
  if (i >= 0) {
    index[i] = toSummary(list);
  } else {
    index.push(toSummary(list));
  }
  index.sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
  saveListsIndex(index);
}

export function removeListIndex(listId: string): void {
  const index = loadListsIndex().filter((s) => s.id !== listId);
  saveListsIndex(index);
}

// ── CRUD ──

export function listLists(): ListSummary[] {
  return loadListsIndex();
}

export function getList(listId: string): List {
  const list = loadList(listId);
  if (!list) throw new Error(`list ${listId} not found`);
  return list;
}

export function createList(name: string, fields: ListField[] = []): List {
  const now = new Date().toISOString();
  const list: List = {
    id: randomUUID(),
    name,
    fields,
    items: [],
    created_at: now,
    updated_at: now,
  };
  saveList(list);
  upsertListIndex(list);
  return list;
}

export function updateListMeta(listId: string, name: string, fields: ListField[]): List {
  const list = getList(listId);
  const kept = new Set(fields.map((f) => f.key));
  for (const item of list.items) {
    for (const key of Object.keys(item.values)) {
      if (!kept.has(key)) delete item.values[key];
    }
  }
  list.name = name;
  list.fields = fields;
  list.updated_at = new Date().toISOString();
  saveList(list);
  upsertListIndex(list);
  return list;
}

export function deleteList(listId: string): void {
  deleteListFile(listId);
  removeListIndex(listId);
}

export function addListItem(listId: string, values: Record<string, unknown>): List {
  const list = getList(listId);
  const now = new Date().toISOString();
  list.items.push({
    id: randomUUID(),
    values: values ?? {},
    linked_todos: [],
    archived: false,
    created_at: now,
    updated_at: now,
  });
  list.updated_at = now;
  saveList(list);
  upsertListIndex(list);
  return list;
}

export function updateListItem(listId: string, item: ListItem): List {
  const list = getList(listId);
  const existing = list.items.find((i) => i.id === item.id);
  if (existing) {
    existing.values = item.values;
    existing.archived = item.archived ?? false;
    existing.updated_at = new Date().toISOString();
  }
  list.updated_at = new Date().toISOString();
  saveList(list);
  upsertListIndex(list);
  return list;
}

export function deleteListItem(listId: string, itemId: string): List {
  const list = getList(listId);
  list.items = list.items.filter((i) => i.id !== itemId);
  list.updated_at = new Date().toISOString();
  saveList(list);
  upsertListIndex(list);
  return list;
}
